### PR TITLE
Issue 16181: Add tests and fixes for Kerberos Ldap

### LINKED ISSUE #16181: KRB5: Add Keytab and Config FAT test for Kerberos Ldap and other to-dos
For #12794

See PR for list of fixes and test updates.

### PR DESCRIPTION
**Pull request by kristip17**
Fixes #16181 

Runtime: Updated `KerberosService` to set the config to original JVM setting if modify is removing the config value. (added clear to helper class `SecureAction`)

Runtime: Clear attribute and search cache when the `<Kerberos/>` config is updated

Runtime: Adjusted some of the error handling and tracing in `ContextManager`

Addressed last batch of review comments from #16171

Added some LeakedPassword detection in the tests

Added several keytab and config tests

Add/expanded simple/ticketcache tests

Added more test comments

Removed unnecessary `setTraceMarkToEndOfDefaultTrace` --- the `updateConfigDynamically` resets log marks.

Swapped trace checks to `findStringsInLogsAndTraceUsingMark`

Customize stop messages allowed per test instead of global list in `CommonBindTest`

### CODE REVIEW

**Comment by kristip17** [comment]
#build

**Review by jvanhill — CHANGES_REQUESTED** [review]
I have a few small changes. I also wonder about the wrongUser.keytab. Is it always going to fail b/c it is the wrong user, or will it fail b/c the credential inside it expired at some point? Or does it not really matter?

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConstants.java [hunk 1]
```diff
@@ -227,4 +227,6 @@
     String SECUREWAY_LDAP_SERVER = "IBM SECUREWAY DIRECTORY SERVER";
 
     String CUSTOM_LDAP_SERVER = "CUSTOM";
+
+    String KERBEROS_UDPATE_MSG = "The Kerberos configuration was updated, refresh the context pool.";
```
[the remark this answers is not in the document]
Comment (kristip17) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConstants.java:231]: `recreateDirContext` takes an errorMessage to log so we don't lose the reason why we're running it. I set it the message in a constant so it'd be easier to keep consistent if we change it in the future.

**Comment by kristip17** [comment]
> I have a few small changes. I also wonder about the wrongUser.keytab. Is it always going to fail b/c it is the wrong user, or will it fail b/c the credential inside it expired at some point? Or does it not really matter?

It is supposed to fail because the user can't be found in the keytab which is generally a valid keytab (versus an empty file or wrong file name) but it looks like I have an empty keytab instead of a populated one. I will fix it in my next round of fat updates -- #16226

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/TicketCacheBindTest.java [hunk 2]
```diff
@@ -57,6 +65,25 @@ public void basicLoginChecks() throws Exception {
...
     }
 
+    /**
+     * Run golden path tests for bindAuthMech=GSSAPI, a valid krbPrincipalName and valid krb5TicketCache
+     *
+     * @throws Exception
+     */
+    @Test
+    @CheckForLeakedPasswords(LdapKerberosUtils.BIND_PASSWORD)
+    @AllowedFFDC("javax.naming.NamingException") // temporary, remove when Issue #16231 is fixed
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/TicketCacheBindTest.java:75]: Do you have a comment in 16231 to come back and remove this?

**Review by jvanhill — APPROVED** [review]

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/TicketCacheBindTest.java (hunk 2 above)
Comment (kristip17) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/TicketCacheBindTest.java:75]: It does-- #16231

**Comment by kristip17** [comment]
#libby

**Comment by kristip17** [comment]
All krb5 ldap tests ran, test failure is a known issue.

**Pull request merged**